presto-tests/src/test/java/com/facebook/presto/tests/TestHistoryBasedRetry.java

```diff
@@ -0,0 +1,178 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tests;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.server.BasicQueryInfo;
+import com.facebook.presto.spi.Plugin;
+import com.facebook.presto.spi.statistics.HistoryBasedPlanStatisticsProvider;
+import com.facebook.presto.testing.InMemoryHistoryBasedPlanStatisticsProvider;
+import com.facebook.presto.testing.QueryRunner;
+import com.facebook.presto.tpch.TpchPlugin;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import static com.facebook.presto.SystemSessionProperties.QUERY_RETRY_LIMIT;
+import static com.facebook.presto.SystemSessionProperties.RESTRICT_HISTORY_BASED_OPTIMIZATION_TO_COMPLEX_QUERY;
+import static com.facebook.presto.SystemSessionProperties.RETRY_QUERY_WITH_HISTORY_BASED_OPTIMIZATION;
+import static com.facebook.presto.SystemSessionProperties.TRACK_HISTORY_BASED_PLAN_STATISTICS;
+import static com.facebook.presto.SystemSessionProperties.TRACK_HISTORY_STATS_FROM_FAILED_QUERIES;
+import static com.facebook.presto.SystemSessionProperties.USE_HISTORY_BASED_PLAN_STATISTICS;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.testng.Assert.assertEquals;
+
+public class TestHistoryBasedRetry
+{
+    private ListeningExecutorService executor;
+
+    @BeforeClass(alwaysRun = true)
+    public void setUp()
+    {
+        executor = MoreExecutors.listeningDecorator(newCachedThreadPool());
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void shutdown()
+    {
+        executor.shutdownNow();
+    }
+
+    @Test
+    public void testQueryRetryWithHBONaturalFail()
+            throws Exception
+    {
+        int totalQueries = 1;
+        int retryLimit = 3;
+        int retryNum = 0;
+
+        String sql = "SELECT o.orderkey, l.partkey, l.mapcol[o.orderkey] FROM (select orderkey, partkey, mapcol " +
+                "FROM (SELECT *, map(array[1], array[2]) mapcol FROM lineitem)) l " +
+                "JOIN orders o ON l.partkey = o.custkey WHERE length(comment) > 10";
+
+        Session session = Session.builder(createSession())
+                .setSystemProperty(QUERY_RETRY_LIMIT, String.valueOf(retryLimit))
```

Comment:
This is not used in HBO retry?

Reply:
@feilong-liu I wanted to set it to a number higher than 1 in one of the tests to ensure the retry count stayed as 1